app/grandchallenge/components/backends/docker.py

```diff
@@ -479,53 +493,61 @@ def _create_images_result(self, *, interface, reader):
             )
 
         return civ
 
-    def _create_json_result(self, *, interface, reader):
-        output_file = Path(safe_join("/output/", interface.relative_path))
+    def _create_json_result(self, *, interface):
+        key = safe_join("/", self.io_prefix, interface.relative_path)
 
         try:
-            with TemporaryDirectory() as tmpdir:
-                temp_file = Path(safe_join(tmpdir, "output.json"))
-                get_file(container=reader, src=output_file, dest=temp_file)
-
-                with open(temp_file, "rb") as file:
-                    result = json.loads(
-                        file.read().decode("utf-8"),
-                        parse_constant=lambda x: None,  # Removes -inf, inf and NaN
-                    )
-        except NotFound:
-            raise ComponentException(f"File {output_file} was not produced")
+            with io.BytesIO() as fileobj:
+                self._s3_client.download_fileobj(
+                    Fileobj=fileobj,
+                    Bucket=settings.COMPONENTS_OUTPUT_BUCKET_NAME,
+                    Key=key,
+                )
+                fileobj.seek(0)
+                result = json.loads(
+                    fileobj.read().decode("utf-8"),
+                    parse_constant=lambda x: None,  # Removes -inf, inf and NaN
+                )
+            civ = interface.create_instance(value=result)
+        except botocore.exceptions.ClientError:
+            raise ComponentException(
+                f"Output file '{interface.relative_path}' was not produced"
+            )
         except JSONDecodeError:
             raise ComponentException(
-                f"The file produced at {output_file} is not valid json"
+                f"The output file '{interface.relative_path}' is not valid json"
             )
-
-        try:
-            civ = interface.create_instance(value=result)
         except ValidationError:
             raise ComponentException(
-                f"The file produced at {output_file} is not valid"
+                f"The output file '{interface.relative_path}' is not valid"
             )
 
         return civ
 
-    def _create_file_result(self, *, interface, reader):
-        output_file = Path(safe_join("/output/", interface.relative_path))
+    def _create_file_result(self, *, interface):
+        key = safe_join("/", self.io_prefix, interface.relative_path)
+
         try:
-            with TemporaryDirectory() as tmpdir:
-                temp_file = Path(safe_join(tmpdir, interface.relative_path))
-                get_file(container=reader, src=output_file, dest=temp_file)
-                with open(temp_file, "rb") as f:
-                    civ = interface.create_instance(fileobj=f)
-        except NotFound:
-            raise ComponentException(f"File {output_file} was not produced")
+            with SpooledTemporaryFile(max_size=MAX_SPOOL_SIZE) as fileobj:
```

Comment:
What happens of the file is over 1GB? Or is that guaranteed to not happen?

Reply:
It then gets written to disk, nice way to keep things fast for small files but still be able to handle larger ones without OOMing:

> This class operates exactly as [TemporaryFile()](https://docs.python.org/3/library/tempfile.html#tempfile.TemporaryFile) does, except that data is spooled in memory until the file size exceeds max_size, or until the file’s fileno() method is called, at which point the contents are written to disk and operation proceeds as with [TemporaryFile()](https://docs.python.org/3/library/tempfile.html#tempfile.TemporaryFile).

https://docs.python.org/3/library/tempfile.html#tempfile.SpooledTemporaryFile